### ISSUE TITLE
Chef 12.1 not allowing control of exit codes with exceptions

### ISSUE THREAD

**Issue by carpnick** [chef/chef#3054]
This ruby block below in Chef 11.* makes chef-client exit with exit code of 250.  In Chef 12 it exits with  1.  The need from a user perspective, is we need to be able to signal a Continuous Delivery system with exit codes out of Chef.  We were depending directly on the exit code in this example, to wait for WinRM or SSH to start running chef again.  The exit code will let us know what action was performed in Chef.  Reboot, Chef run success, etc.
```
ruby_block "ExitNow" do
  block do
     puts "Stopping chef run for an immediate reboot";       
     exit 250;  
  end
  action :run
end
```

Looks like Chef 11's implementation of [Converge](https://github.com/chef/chef/blob/11-stable/lib/chef/client.rb#L341) would raise the exception where in Chef 12 it [catches it](https://github.com/chef/chef/blob/12-stable/lib/chef/client.rb#L324) and returns it to the [run method] (https://github.com/chef/chef/blob/12-stable/lib/chef/client.rb#L448) and then [wraps it](https://github.com/chef/chef/blob/12-stable/lib/chef/client.rb#L461).

I think the potential fix is inside the code where it is wrapping it:
```
if converge_error.class == SystemExit
  raise
end
```

**Comment by carpnick** [comment]
Other option is to have the [chef reboot] (https://docs.chef.io/resource_reboot.html) resource have an option on it to reboot with a specific exit code.

I would rather have this option, but I think it would cause the above comment code change anyways.

**Comment by pburkholder** [comment]
Adding an explicit 'Exit' resource to Chef language would be a solution too.

**Comment by jdmundrawala** [comment]
cc @tyler-ball @mcquin

**Comment by tyler-ball** [comment]
We specifically catch and wrap the exception so we can run the audit phase independently of the converge phase.  So I don't want to implement a re-raise on SystemExit error because that would skip the audit phase.  Is using `exit \d` a typical pattern for flow control?  It doesn't seem to me like it should be.

Thats why I would rather have the reboot resource updated to take in an optional exit code that defaults to 0.

Other opinions?

**Comment by carpnick** [comment]
Updating the reboot resource immediately solves our problem.  

However, I can see other use cases where a user might want to exit chef mid chef-run and not only because of a reboot to signal another system.  Network unavailable, system out of space, etc.

Basically - the question being asked here really, is how do I signal other systems outside of a chef-run the status of the chef run?  The easiest use case is to use exit codes.  Others involve reading the output of Chef and parsing.  But the latter options seems clunky and version dependent to me.

**Comment by btm** [comment]
I think the temporary fix for #3078 in #3090 fixes this. It will be released in 12.1.2

**Comment by carpnick** [comment]
Thanks.
